File: swift/test/test_csv.swift
/**
 * test_csv.swift - Test MPC designation conversion against CSV test data
 *
 * Usage: test_csv <csv_file>
 *
 * CSV format: unpacked,packed (one pair per line)
 */

import Foundation

var totalTests = 0
var passedTests = 0
var failedTests = 0

func runTests(csvPath: String) {
    guard let fileHandle = FileHandle(forReadingAtPath: csvPath) else {
        fputs("Error: Cannot open file: \(csvPath)\n", stderr)
        exit(1)
    }

    guard let data = try? fileHandle.readToEnd(),
          let content = String(data: data, encoding: .utf8) else {
        fputs("Error: Cannot read file: \(csvPath)\n", stderr)
        exit(1)
    }

    let startTime = Date()
    let lines = content.components(separatedBy: .newlines)
    var isFirstLine = true

    for line in lines {
        let trimmed = line.trimmingCharacters(in: .whitespaces)

        // Skip header line
        if isFirstLine {
            isFirstLine = false
            continue
        }

        // Skip empty lines and comments
        if trimmed.isEmpty || trimmed.hasPrefix("#") {
            continue
        }

        // Parse CSV line
        let parts = trimmed.components(separatedBy: ",")
        guard parts.count >= 2 else {
            continue
        }

        let unpacked = parts[0]
        let expectedPacked = parts[1]

        totalTests += 1

        // Test unpacked -> packed (one direction only, matching Python test)
        do {
            let gotPacked = try convertSimple(unpacked)
            if gotPacked != expectedPacked {
                failedTests += 1
                if failedTests <= 100 {
                    print("FAIL: convert('\(unpacked)') = '\(gotPacked)', expected '\(expectedPacked)'")
                }
                continue
            }
        } catch {
            failedTests += 1
            if failedTests <= 100 {
                print("FAIL: convert('\(unpacked)') threw error: \(error)")
            }
            continue
        }

        passedTests += 1

        // Progress indicator
        if totalTests % 100000 == 0 {
            print("Processed \(totalTests) entries...")
        }
    }

    let elapsed = Date().timeIntervalSince(startTime)
    let elapsedMs = Int(elapsed * 1000)
    let rate = totalTests > 0 ? Double(totalTests) / elapsed : 0

    print("")
    print("=== Test Results ===")
    print("Total:  \(totalTests)")
    print("Passed: \(passedTests)")
    print("Failed: \(failedTests)")
    print("Time:   \(elapsedMs)ms (\(String(format: "%.1f", rate)) entries/sec)")
}

// Main entry point for test_csv
func testCSVMain() {
    let args = Array(CommandLine.arguments.dropFirst())

    if args.isEmpty {
        fputs("Usage: test_csv <csv_file>\n", stderr)
        exit(1)
    }

    runTests(csvPath: args[0])
    exit(failedTests > 0 ? 1 : 0)
}

File: haskell/test/TestCsv.hs
{-|
Module      : TestCsv
Description : Test MPC designation against CSV test data
-}

module Main where

import System.Environment (getArgs)
import System.Exit (exitFailure, exitSuccess)
import System.IO (hPutStrLn, stderr)
import Data.Time.Clock (getCurrentTime, diffUTCTime)
import Control.DeepSeq (deepseq)
import MPCDesignation

main :: IO ()
main = do
    args <- getArgs
    case args of
        [csvFile] -> runTests csvFile
        _ -> do
            hPutStrLn stderr "Usage: test_csv <csv_file>"
            exitFailure

runTests :: FilePath -> IO ()
runTests csvFile = do
    contents <- readFile csvFile
    let allLines = lines contents
        dataLines = drop 1 allLines  -- Skip header
        testCases = map parseLine dataLines

    putStrLn $ "Loaded " ++ show (length testCases) ++ " test cases"
    putStrLn ""

    -- Phase 1: Pack (unpacked -> packed)
    putStrLn "=== Phase 1: Pack (unpacked -> packed) ==="
    startTime <- getCurrentTime

    let packResults = map testPack testCases
        (packPassed, packFailed) = countResults packResults

    -- Force evaluation
    packPassed `deepseq` packFailed `deepseq` return ()

    endTime <- getCurrentTime
    let elapsed = realToFrac (diffUTCTime endTime startTime) * 1000 :: Double
        rate = fromIntegral (length testCases) / (elapsed / 1000)

    putStrLn $ "Passed: " ++ show packPassed
    putStrLn $ "Failed: " ++ show packFailed
    putStrLn $ "Time:   " ++ show (round elapsed :: Int) ++ "ms (" ++ show (round rate :: Int) ++ ".0 entries/sec)"
    putStrLn ""

    if packFailed > 0
        then exitFailure
        else exitSuccess

parseLine :: String -> (String, String)
parseLine line =
    let (unpacked, rest) = break (== ',') line
        packed = drop 1 rest
    in (unpacked, packed)

testPack :: (String, String) -> Bool
testPack (unpacked, expected) =
    case pack unpacked of
        Right result -> result == expected
        Left _ -> False

countResults :: [Bool] -> (Int, Int)
countResults = foldr count (0, 0)
  where
    count True (p, f) = (p + 1, f)
    count False (p, f) = (p, f + 1)
